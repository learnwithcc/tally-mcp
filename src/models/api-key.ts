import { z } from 'zod';

/**
 * API Key Status enum
 */
export enum ApiKeyStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  REVOKED = 'revoked',
  EXPIRED = 'expired'
}

/**
 * API Key Permission Scopes
 */
export enum ApiKeyScope {
  READ = 'read',
  WRITE = 'write',
  ADMIN = 'admin',
  FORMS_READ = 'forms:read',
  FORMS_WRITE = 'forms:write',
  SUBMISSIONS_READ = 'submissions:read',
  SUBMISSIONS_WRITE = 'submissions:write',
  WORKSPACES_READ = 'workspaces:read',
  WORKSPACES_WRITE = 'workspaces:write'
}

/**
 * Schema for creating a new API key
 */
export const CreateApiKeySchema = z.object({
  name: z.string().min(1, 'API key name is required').max(100, 'Name too long'),
  description: z.string().optional(),
  scopes: z.array(z.nativeEnum(ApiKeyScope)).min(1, 'At least one scope is required'),
  expiresAt: z.date().optional(),
  maxUsage: z.number().positive().optional(),
  ipWhitelist: z.array(z.string().ip()).optional(),
  metadata: z.record(z.string()).optional()
});

/**
 * Schema for updating an API key
 */
export const UpdateApiKeySchema = z.object({
  name: z.string().min(1).max(100).optional(),
  description: z.string().optional(),
  scopes: z.array(z.nativeEnum(ApiKeyScope)).optional(),
  status: z.nativeEnum(ApiKeyStatus).optional(),
  expiresAt: z.date().optional(),
  maxUsage: z.number().positive().optional(),
  ipWhitelist: z.array(z.string().ip()).optional(),
  metadata: z.record(z.string()).optional()
});

/**
 * Schema for API key validation
 */
export const ValidateApiKeySchema = z.object({
  key: z.string().min(1, 'API key is required'),
  ipAddress: z.string().ip().optional(),
  userAgent: z.string().optional(),
  endpoint: z.string().optional()
});

/**
 * API Key interface
 */
export interface ApiKey {
  id: string;
  keyHash: string; // Hashed version of the key for storage
  name: string;
  description?: string;
  scopes: ApiKeyScope[];
  status: ApiKeyStatus;
  createdAt: Date;
  updatedAt: Date;
  expiresAt?: Date;
  lastUsedAt?: Date;
  usageCount: number;
  maxUsage?: number;
  ipWhitelist?: string[];
  metadata?: Record<string, string>;
  createdBy?: string;
}

/**
 * API Key with plain text key (only returned on creation)
 */
export interface ApiKeyWithSecret extends Omit<ApiKey, 'keyHash'> {
  key: string; // Plain text key, only shown once
  keyHash: string;
}

/**
 * API Key usage log entry
 */
export interface ApiKeyUsageLog {
  id: string;
  keyId: string;
  timestamp: Date;
  ipAddress?: string;
  userAgent?: string;
  endpoint?: string;
  success: boolean;
  errorReason?: string;
  responseTime?: number;
}

/**
 * Schema for API key usage logging
 */
export const ApiKeyUsageLogSchema = z.object({
  keyId: z.string().uuid(),
  ipAddress: z.string().ip().optional(),
  userAgent: z.string().optional(),
  endpoint: z.string().optional(),
  success: z.boolean(),
  errorReason: z.string().optional(),
  responseTime: z.number().positive().optional()
});

/**
 * API Key rotation configuration
 */
export interface ApiKeyRotationConfig {
  id: string;
  keyId: string;
  rotationIntervalDays: number;
  notifyDaysBefore: number;
  autoRotate: boolean;
  lastRotatedAt?: Date;
  nextRotationAt: Date;
  isActive: boolean;
}

/**
 * Schema for API key rotation configuration
 */
export const ApiKeyRotationConfigSchema = z.object({
  rotationIntervalDays: z.number().min(1).max(365),
  notifyDaysBefore: z.number().min(0).max(30),
  autoRotate: z.boolean(),
  isActive: z.boolean().default(true)
});

/**
 * Type definitions for API key operations
 */
export type CreateApiKeyInput = z.infer<typeof CreateApiKeySchema>;
export type UpdateApiKeyInput = z.infer<typeof UpdateApiKeySchema>;
export type ValidateApiKeyInput = z.infer<typeof ValidateApiKeySchema>;
export type ApiKeyUsageLogInput = z.infer<typeof ApiKeyUsageLogSchema>;
export type ApiKeyRotationConfigInput = z.infer<typeof ApiKeyRotationConfigSchema>;

/**
 * API Key validation result
 */
export interface ApiKeyValidationResult {
  isValid: boolean;
  apiKey?: ApiKey;
  errorReason?: string;
  remainingUsage?: number | undefined;
  expiresIn?: number | undefined; // seconds until expiration
}

/**
 * API Key generation options
 */
export interface ApiKeyGenerationOptions {
  length?: number;
  prefix?: string;
  includeChecksum?: boolean;
}

/**
 * Constants for API key management
 */
export const API_KEY_CONSTANTS = {
  DEFAULT_KEY_LENGTH: 32,
  DEFAULT_PREFIX: 'tally_',
  MIN_KEY_LENGTH: 16,
  MAX_KEY_LENGTH: 64,
  HASH_ALGORITHM: 'sha256',
  ENCRYPTION_ALGORITHM: 'aes-256-gcm',
  DEFAULT_EXPIRY_DAYS: 90,
  MAX_USAGE_LOGS_PER_KEY: 1000,
  RATE_LIMIT_WINDOW_MS: 60000, // 1 minute
  RATE_LIMIT_MAX_REQUESTS: 100
} as const; 