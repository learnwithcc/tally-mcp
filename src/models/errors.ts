/**
 * Standardized Error Classes for Tally API Client
 * 
 * These error classes provide structured error handling for different
 * HTTP status codes and API failure scenarios.
 */

/**
 * Base API error class that all other API errors extend
 */
export abstract class TallyApiError extends Error {
  public readonly isRetryable: boolean;
  public readonly statusCode?: number;
  public readonly statusText?: string;
  public readonly headers?: Record<string, string>;
  public readonly requestId?: string;
  public readonly timestamp: Date;
  public readonly cause?: Error;

  constructor(
    message: string,
    statusCode?: number,
    statusText?: string,
    headers?: Record<string, string>,
    isRetryable: boolean = false,
    cause?: Error
  ) {
    super(message);
    this.name = this.constructor.name;
    if (statusCode !== undefined) this.statusCode = statusCode;
    if (statusText !== undefined) this.statusText = statusText;
    if (headers !== undefined) this.headers = headers;
    this.isRetryable = isRetryable;
    this.timestamp = new Date();
    if (headers?.['x-request-id']) this.requestId = headers['x-request-id'];
    else if (headers?.['request-id']) this.requestId = headers['request-id'];
    if (cause !== undefined) this.cause = cause;
  }

  /**
   * Get a user-friendly error message
   */
  public getUserMessage(): string {
    return this.message;
  }

  /**
   * Get detailed error information for debugging
   */
  public getDebugInfo(): Record<string, any> {
    return {
      name: this.name,
      message: this.message,
      statusCode: this.statusCode,
      statusText: this.statusText,
      requestId: this.requestId,
      timestamp: this.timestamp.toISOString(),
      isRetryable: this.isRetryable,
      headers: this.headers,
    };
  }
}

/**
 * Authentication Error (401/403)
 * Thrown when authentication fails or token is invalid
 */
export class AuthenticationError extends TallyApiError {
  constructor(
    message: string = 'Authentication failed',
    statusCode?: number,
    statusText?: string,
    headers?: Record<string, string>,
    cause?: Error
  ) {
    super(message, statusCode, statusText, headers, false, cause);
  }

  public override getUserMessage(): string {
    if (this.statusCode === 401) {
      return 'Your session has expired. Please log in again.';
    }
    if (this.statusCode === 403) {
      return 'You do not have permission to access this resource.';
    }
    return 'Authentication failed. Please check your credentials.';
  }
}

/**
 * Bad Request Error (400)
 * Thrown when the request is malformed or invalid
 */
export class BadRequestError extends TallyApiError {
  public readonly validationErrors?: Array<{
    field?: string;
    message: string;
    code?: string;
  }>;

  constructor(
    message: string = 'Bad request',
    statusCode: number = 400,
    statusText?: string,
    headers?: Record<string, string>,
    validationErrors?: Array<{ field?: string; message: string; code?: string }>,
    cause?: Error
  ) {
    super(message, statusCode, statusText, headers, false, cause);
    if (validationErrors !== undefined) this.validationErrors = validationErrors;
  }

  public override getUserMessage(): string {
    if (this.validationErrors && this.validationErrors.length > 0) {
      const fieldErrors = this.validationErrors
        .filter(error => error.field)
        .map(error => `${error.field}: ${error.message}`)
        .join(', ');
      
      if (fieldErrors) {
        return `Validation failed: ${fieldErrors}`;
      }
    }
    return 'The request contains invalid data. Please check your input and try again.';
  }
}

/**
 * Not Found Error (404)
 * Thrown when the requested resource is not found
 */
export class NotFoundError extends TallyApiError {
  constructor(
    message: string = 'Resource not found',
    statusCode: number = 404,
    statusText?: string,
    headers?: Record<string, string>,
    cause?: Error
  ) {
    super(message, statusCode, statusText, headers, false, cause);
  }

  public override getUserMessage(): string {
    return 'The requested resource was not found.';
  }
}

/**
 * Rate Limit Error (429)
 * Thrown when rate limits are exceeded
 */
export class RateLimitError extends TallyApiError {
  public readonly retryAfter?: number;
  public readonly limit?: number;
  public readonly remaining?: number;
  public readonly resetTime?: Date;

  constructor(
    message: string = 'Rate limit exceeded',
    statusCode: number = 429,
    statusText?: string,
    headers?: Record<string, string>,
    cause?: Error
  ) {
    super(message, statusCode, statusText, headers, true, cause);
    
    // Parse rate limit headers
    if (headers) {
      if (headers['retry-after']) {
        const retryAfterParsed = parseInt(headers['retry-after'], 10);
        if (!isNaN(retryAfterParsed)) this.retryAfter = retryAfterParsed;
      }
      if (headers['x-ratelimit-limit']) {
        const limitParsed = parseInt(headers['x-ratelimit-limit'], 10);
        if (!isNaN(limitParsed)) this.limit = limitParsed;
      }
      if (headers['x-ratelimit-remaining']) {
        const remainingParsed = parseInt(headers['x-ratelimit-remaining'], 10);
        if (!isNaN(remainingParsed)) this.remaining = remainingParsed;
      }
      if (headers['x-ratelimit-reset']) {
        const resetTimeParsed = parseInt(headers['x-ratelimit-reset'], 10);
        if (!isNaN(resetTimeParsed)) this.resetTime = new Date(resetTimeParsed * 1000);
      }
    }
  }

  public override getUserMessage(): string {
    if (this.retryAfter) {
      return `Rate limit exceeded. Please wait ${this.retryAfter} seconds before trying again.`;
    }
    if (this.resetTime) {
      return `Rate limit exceeded. Limit resets at ${this.resetTime.toLocaleTimeString()}.`;
    }
    return 'Rate limit exceeded. Please try again later.';
  }
}

/**
 * Server Error (500+)
 * Thrown when the server encounters an internal error
 */
export class ServerError extends TallyApiError {
  constructor(
    message: string = 'Internal server error',
    statusCode: number = 500,
    statusText?: string,
    headers?: Record<string, string>,
    cause?: Error
  ) {
    // Server errors are typically retryable
    super(message, statusCode, statusText, headers, true, cause);
  }

  public override getUserMessage(): string {
    if (this.statusCode === 502) {
      return 'Service temporarily unavailable. Please try again later.';
    }
    if (this.statusCode === 503) {
      return 'Service is currently under maintenance. Please try again later.';
    }
    if (this.statusCode === 504) {
      return 'Request timed out. Please try again.';
    }
    return 'A server error occurred. Please try again later.';
  }
}

/**
 * Network Error
 * Thrown when network connectivity issues occur
 */
export class NetworkError extends TallyApiError {
  constructor(
    message: string = 'Network error',
    cause?: Error
  ) {
    super(message, undefined, undefined, undefined, true, cause);
  }

  public override getUserMessage(): string {
    return 'Network connection failed. Please check your internet connection and try again.';
  }
}

/**
 * Timeout Error
 * Thrown when requests exceed the configured timeout
 */
export class TimeoutError extends TallyApiError {
  constructor(
    message: string = 'Request timeout',
    cause?: Error
  ) {
    super(message, undefined, undefined, undefined, true, cause);
  }

  public override getUserMessage(): string {
    return 'Request timed out. Please try again.';
  }
}

/**
 * Factory function to create appropriate error based on HTTP status code
 */
export function createErrorFromResponse(
  statusCode: number,
  statusText: string,
  headers: Record<string, string>,
  responseData?: any,
  originalError?: Error
): TallyApiError {
  const message = responseData?.message || responseData?.error || statusText || 'Unknown error';
  
  switch (statusCode) {
    case 400:
      return new BadRequestError(
        message,
        statusCode,
        statusText,
        headers,
        responseData?.errors || responseData?.validation_errors,
        originalError
      );
    
    case 401:
    case 403:
      return new AuthenticationError(message, statusCode, statusText, headers, originalError);
    
    case 404:
      return new NotFoundError(message, statusCode, statusText, headers, originalError);
    
    case 429:
      return new RateLimitError(message, statusCode, statusText, headers, originalError);
    
    default:
      if (statusCode >= 500) {
        return new ServerError(message, statusCode, statusText, headers, originalError);
      }
      
      // For other 4xx errors, use BadRequestError as fallback
      if (statusCode >= 400 && statusCode < 500) {
        return new BadRequestError(message, statusCode, statusText, headers, undefined, originalError);
      }
      
      // Fallback for unexpected status codes
      return new ServerError(message, statusCode, statusText, headers, originalError);
  }
} 