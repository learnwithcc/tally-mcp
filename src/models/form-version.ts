import { FormConfig } from './form-config';

/**
 * Represents a snapshot of a form at a specific point in time.
 */
export interface FormVersion {
  /**
   * Version number (integer, increments with each change)
   */
  version: number;

  /**
   * The form configuration for this version
   */
  formConfig: FormConfig;

  /**
   * ISO string timestamp of when this version was created
   */
  createdAt: string;

  /**
   * A brief description of the changes made in this version
   */
  changeSummary?: string | undefined;
}

/**
 * Manages the version history of a form.
 */
export class FormVersionManager {
  private history: FormVersion[] = [];

  constructor(initialFormConfig: FormConfig) {
    this.addVersion(initialFormConfig, 'Initial version');
  }

  /**
   * Adds a new version to the history.
   * @param formConfig The new form configuration.
   * @param changeSummary A summary of the changes.
   * @returns The newly created version.
   */
  public addVersion(formConfig: FormConfig, changeSummary?: string): FormVersion {
    const versionNumber = formConfig.metadata?.version || this.history.length + 1;
    
    const existingVersion = this.getVersion(versionNumber);
    if (existingVersion) {
      existingVersion.changeSummary = changeSummary;
      existingVersion.formConfig = JSON.parse(JSON.stringify(formConfig));
      return existingVersion;
    }

    const newVersion: FormVersion = {
      version: versionNumber,
      formConfig: JSON.parse(JSON.stringify(formConfig)), // Deep copy
      createdAt: formConfig.metadata?.updatedAt || new Date().toISOString(),
      changeSummary: changeSummary,
    };
    
    this.history.push(newVersion);
    this.history.sort((a, b) => a.version - b.version); // Keep sorted
    return newVersion;
  }

  /**
   * Retrieves a specific version by its number.
   * @param versionNumber The version to retrieve.
   * @returns The form version, or undefined if not found.
   */
  public getVersion(versionNumber: number): FormVersion | undefined {
    return this.history.find(v => v.version === versionNumber);
  }

  /**
   * Gets the latest version of the form.
   * @returns The current form version, or undefined if history is empty.
   */
  public getCurrentVersion(): FormVersion | undefined {
    if (this.history.length === 0) {
      return undefined;
    }
    return this.history[this.history.length - 1];
  }

  /**
   * Rolls back to a previous version of the form.
   * @param versionNumber The version number to roll back to.
   * @returns The form configuration of the target version, or undefined if the version doesn't exist.
   */
  public rollbackTo(versionNumber: number): FormConfig | undefined {
    const targetVersion = this.getVersion(versionNumber);
    if (!targetVersion) {
      return undefined;
    }

    // Remove all versions after the target version
    this.history = this.history.filter(v => v.version <= versionNumber);
    
    return this.getCurrentVersion()?.formConfig;
  }

  /**
   * Returns the entire version history.
   */
  public getHistory(): FormVersion[] {
    return this.history;
  }
} 