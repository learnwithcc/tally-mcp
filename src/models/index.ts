/**
 * Data Models and Schemas
 * 
 * This directory contains TypeScript interfaces and Zod schemas for:
 * - Form configurations and structures
 * - Question types and validation rules
 * - API request/response types
 * - Tally.so data models
 */

// Export error classes for API client
export {
  TallyApiError,
  AuthenticationError,
  BadRequestError,
  NotFoundError,
  RateLimitError,
  ServerError,
  NetworkError,
  TimeoutError,
  createErrorFromResponse,
} from './errors';

// Export Tally API schemas and types
export * from './tally-schemas';

// Export form configuration interfaces and types
export * from './form-config'; 