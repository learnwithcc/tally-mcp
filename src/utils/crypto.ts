import crypto from 'crypto';
import { API_KEY_CONSTANTS, ApiKeyGenerationOptions } from '../models/api-key';

/**
 * Encryption result interface
 */
export interface EncryptionResult {
  encrypted: string;
  iv: string;
  tag: string;
}

/**
 * Decryption input interface
 */
export interface DecryptionInput {
  encrypted: string;
  iv: string;
  tag: string;
}

/**
 * Crypto utilities for API key management
 */
export class CryptoUtils {
  private static readonly algorithm = API_KEY_CONSTANTS.ENCRYPTION_ALGORITHM;
  private static readonly hashAlgorithm = API_KEY_CONSTANTS.HASH_ALGORITHM;

  /**
   * Get encryption key from environment variable
   */
  private static getEncryptionKey(): Buffer {
    const key = process.env.API_KEY_ENCRYPTION_KEY;
    if (!key) {
      throw new Error('API_KEY_ENCRYPTION_KEY environment variable is required');
    }
    
    // If key is shorter than 32 bytes, pad it with zeros
    const keyBuffer = Buffer.from(key, 'utf8');
    if (keyBuffer.length < 32) {
      const paddedKey = Buffer.alloc(32);
      keyBuffer.copy(paddedKey);
      return paddedKey;
    }
    
    // If key is longer than 32 bytes, truncate it
    return keyBuffer.slice(0, 32);
  }

  /**
   * Generate a cryptographically secure API key
   */
  public static generateApiKey(options: ApiKeyGenerationOptions = {}): string {
    const {
      length = API_KEY_CONSTANTS.DEFAULT_KEY_LENGTH,
      prefix = API_KEY_CONSTANTS.DEFAULT_PREFIX,
    } = options;

    if (length < API_KEY_CONSTANTS.MIN_KEY_LENGTH || length > API_KEY_CONSTANTS.MAX_KEY_LENGTH) {
      throw new Error(`Key length must be between ${API_KEY_CONSTANTS.MIN_KEY_LENGTH} and ${API_KEY_CONSTANTS.MAX_KEY_LENGTH}`);
    }

    const randomPart = crypto.randomBytes(length).toString('hex');
    const key = `${prefix}${randomPart}`;
    const checksum = this.calculateChecksum(key);
    
    return `${key}_${checksum}`;
  }

  /**
   * Calculate checksum for API key validation
   */
  private static calculateChecksum(data: string): string {
    return crypto.createHash('sha256').update(data).digest('hex').substring(0, 8);
  }

  /**
   * Validate API key format and checksum
   */
  public static validateKeyFormat(key: string): boolean {
    const parts = key.split('_');
    if (parts.length < 2) {
      return false;
    }
    
    const checksum = parts.pop();
    const keyWithoutChecksum = parts.join('_');
    
    return this.calculateChecksum(keyWithoutChecksum) === checksum;
  }

  /**
   * Hash an API key for secure storage
   */
  public static hashApiKey(key: string): string {
    return crypto
      .createHash(this.hashAlgorithm)
      .update(key)
      .digest('hex');
  }

  /**
   * Encrypt sensitive data
   */
  public static encrypt(plaintext: string): EncryptionResult {
    const key = this.getEncryptionKey();
    const iv = crypto.randomBytes(16);
    
    const cipher = crypto.createCipheriv(this.algorithm, key, iv);
    
    let encrypted = cipher.update(plaintext, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    
    const tag = (cipher as any).getAuthTag ? (cipher as any).getAuthTag() : Buffer.alloc(0);

    return {
      encrypted,
      iv: iv.toString('hex'),
      tag: tag.toString('hex')
    };
  }

  /**
   * Decrypt sensitive data
   */
  public static decrypt(data: DecryptionInput): string {
    const key = this.getEncryptionKey();
    const iv = Buffer.from(data.iv, 'hex');
    const tag = Buffer.from(data.tag, 'hex');
    
    const decipher = crypto.createDecipheriv(this.algorithm, key, iv);
    
    if ((decipher as any).setAuthTag && data.tag) {
      (decipher as any).setAuthTag(tag);
    }
    
    let decrypted = decipher.update(data.encrypted, 'hex', 'utf8');
    decrypted += decipher.final('utf8');
    
    return decrypted;
  }

  /**
   * Generate a secure random salt
   */
  public static generateSalt(length: number = 16): string {
    return crypto.randomBytes(length).toString('hex');
  }

  /**
   * Create HMAC signature
   */
  public static createHmac(data: string, secret: string): string {
    return crypto
      .createHmac('sha256', secret)
      .update(data)
      .digest('hex');
  }

  /**
   * Verify HMAC signature
   */
  public static verifyHmac(data: string, signature: string, secret: string): boolean {
    const expectedSignature = this.createHmac(data, secret);
    try {
      return crypto.timingSafeEqual(
        Buffer.from(signature, 'hex'),
        Buffer.from(expectedSignature, 'hex')
      );
    } catch {
      // If buffers have different lengths, timingSafeEqual throws an error
      return false;
    }
  }

  /**
   * Generate a unique identifier
   */
  public static generateId(): string {
    return crypto.randomUUID();
  }

  /**
   * Create a secure hash with salt
   */
  public static hashWithSalt(data: string, salt: string): string {
    return crypto
      .createHash('sha256')
      .update(data + salt)
      .digest('hex');
  }

  /**
   * Generate a JWT-like token for temporary authentication
   */
  public static generateTempToken(payload: Record<string, any>, expiresInSeconds: number = 3600): string {
    const header = {
      typ: 'JWT',
      alg: 'HS256'
    };

    const now = Math.floor(Date.now() / 1000);
    const tokenPayload = {
      ...payload,
      iat: now,
      exp: now + expiresInSeconds
    };

    const encodedHeader = Buffer.from(JSON.stringify(header)).toString('base64url');
    const encodedPayload = Buffer.from(JSON.stringify(tokenPayload)).toString('base64url');
    
    const secret = process.env.JWT_SECRET || 'fallback-secret';
    const signature = this.createHmac(`${encodedHeader}.${encodedPayload}`, secret);
    const encodedSignature = Buffer.from(signature, 'hex').toString('base64url');

    return `${encodedHeader}.${encodedPayload}.${encodedSignature}`;
  }

  /**
   * Verify and decode a temporary token
   */
  public static verifyTempToken(token: string): { valid: boolean; payload?: any; error?: string } {
    try {
      const [encodedHeader, encodedPayload, encodedSignature] = token.split('.');
      
      if (!encodedHeader || !encodedPayload || !encodedSignature) {
        return { valid: false, error: 'Invalid token format' };
      }

      const secret = process.env.JWT_SECRET || 'fallback-secret';
      const expectedSignature = this.createHmac(`${encodedHeader}.${encodedPayload}`, secret);
      const expectedEncodedSignature = Buffer.from(expectedSignature, 'hex').toString('base64url');

      if (encodedSignature !== expectedEncodedSignature) {
        return { valid: false, error: 'Invalid signature' };
      }

      const payload = JSON.parse(Buffer.from(encodedPayload, 'base64url').toString('utf8'));
      const now = Math.floor(Date.now() / 1000);

      if (payload.exp && payload.exp < now) {
        return { valid: false, error: 'Token expired' };
      }

      return { valid: true, payload };
    } catch (error) {
      return { valid: false, error: 'Token parsing failed' };
    }
  }

  /**
   * Mask sensitive data for logging
   */
  public static maskSensitiveData(data: string, visibleChars: number = 4): string {
    if (data.length <= visibleChars * 2) {
      return '*'.repeat(data.length);
    }
    
    const start = data.slice(0, visibleChars);
    const end = data.slice(-visibleChars);
    const middle = '*'.repeat(data.length - (visibleChars * 2));
    
    return `${start}${middle}${end}`;
  }
} 