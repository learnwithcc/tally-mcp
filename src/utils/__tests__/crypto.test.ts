import crypto from 'crypto';
import { CryptoUtils, EncryptionResult } from '../crypto';
import { API_KEY_CONSTANTS } from '../../models/api-key';
import { cryptoTestData, temporaryTokenTestCases } from './fixtures';

describe('CryptoUtils', () => {
  const originalEnv = process.env;
  const { plainText, encryptionKey, hmacKey, apiKeyPrefix } = cryptoTestData;

  beforeEach(() => {
    jest.resetModules();
    process.env = { ...originalEnv };
  });

  afterAll(() => {
    process.env = originalEnv;
  });

  describe('API Key Generation and Validation', () => {
    it('should generate a valid API key with default options', () => {
      const key = CryptoUtils.generateApiKey();
      expect(CryptoUtils.validateKeyFormat(key)).toBe(true);
      expect(key).toMatch(/^tally_/);
      expect(key).toMatchSnapshot();
    });

    it('should generate a valid API key with a custom prefix', () => {
      const key = CryptoUtils.generateApiKey({ prefix: apiKeyPrefix });
      expect(CryptoUtils.validateKeyFormat(key)).toBe(true);
      expect(key).toMatch(new RegExp(`^${apiKeyPrefix}`));
      expect(key).toMatchSnapshot();
    });

    it('should invalidate a key with a wrong checksum', () => {
        const key = CryptoUtils.generateApiKey();
        const parts = key.split('_');
        const tamperedKey = `${parts[0]}_wrongchecksum`;
        expect(CryptoUtils.validateKeyFormat(tamperedKey)).toBe(false);
    });

    test.each([
      [10],
      [100],
    ])('should throw an error for invalid key length: %s', (length) => {
      expect(() => CryptoUtils.generateApiKey({ length })).toThrow('Key length must be between');
    });
  });

  describe('Hashing', () => {
    it('should hash an API key consistently', () => {
      const hash1 = CryptoUtils.hashApiKey('some-key');
      const hash2 = CryptoUtils.hashApiKey('some-key');
      expect(hash1).toBe(hash2);
      expect(hash1).toMatchSnapshot();
    });

    it('should produce different hashes for different keys', () => {
      const hash1 = CryptoUtils.hashApiKey('some-key');
      const hash2 = CryptoUtils.hashApiKey('another-key');
      expect(hash1).not.toBe(hash2);
    });

    it('should hash with a salt', () => {
      const data = 'password';
      const salt = CryptoUtils.generateSalt();
      const hash = CryptoUtils.hashWithSalt(data, salt);
      expect(hash).toBe(CryptoUtils.hashWithSalt(data, salt));
      expect(hash).not.toBe(CryptoUtils.hashWithSalt(data, CryptoUtils.generateSalt()));
    });
  });

  describe('Encryption and Decryption', () => {
    beforeEach(() => {
        process.env.API_KEY_ENCRYPTION_KEY = encryptionKey;
    });

    afterEach(() => {
        delete process.env.API_KEY_ENCRYPTION_KEY;
    });

    it('should encrypt and decrypt data successfully', () => {
      const encrypted = CryptoUtils.encrypt(plainText);
      const decrypted = CryptoUtils.decrypt(encrypted);
      expect(decrypted).toBe(plainText);
      expect(encrypted).not.toBe(plainText);
      expect(encrypted).toMatchSnapshot();
    });

    it('should throw error if encryption key is missing', () => {
        delete process.env.API_KEY_ENCRYPTION_KEY;
        expect(() => CryptoUtils.encrypt(plainText)).toThrow('API_KEY_ENCRYPTION_KEY environment variable is required');
    });

    it('should pad a short encryption key', () => {
        process.env.API_KEY_ENCRYPTION_KEY = 'short-key';
        const plaintext = 'test';
        const encrypted = CryptoUtils.encrypt(plaintext);
        expect(CryptoUtils.decrypt(encrypted)).toBe(plaintext);
    });

    it('should truncate a long encryption key', () => {
        process.env.API_KEY_ENCRYPTION_KEY = 'this-is-a-very-long-key-that-will-be-truncated-for-sure';
        const plaintext = 'test';
        const encrypted = CryptoUtils.encrypt(plaintext);
        expect(CryptoUtils.decrypt(encrypted)).toBe(plaintext);
    });
  });

  describe('HMAC Signature', () => {
    it('should create and verify an HMAC signature', () => {
      const signature = CryptoUtils.createHmac(plainText, hmacKey);
      expect(CryptoUtils.verifyHmac(plainText, signature, hmacKey)).toBe(true);
      expect(signature).toMatchSnapshot();
    });

    it('should fail verification for a wrong signature or data', () => {
      const signature = CryptoUtils.createHmac(plainText, hmacKey);
      expect(CryptoUtils.verifyHmac(plainText, 'wrong-signature', hmacKey)).toBe(false);
      expect(CryptoUtils.verifyHmac('wrong-data', signature, hmacKey)).toBe(false);
    });
  });

  describe('Temporary Token', () => {
    beforeEach(() => {
        process.env.JWT_SECRET = hmacKey;
    });

    afterEach(() => {
        delete process.env.JWT_SECRET;
    });

    it('should generate and verify a valid temporary token', () => {
      const { payload, expiresIn } = temporaryTokenTestCases.valid;
      const token = CryptoUtils.generateTempToken(payload, parseInt(expiresIn));
      const verified = CryptoUtils.verifyTempToken(token);
      expect(verified.valid).toBe(true);
      expect(verified.payload).toMatchObject(payload);
    });

    it('should fail verification for a tampered token', () => {
        const { payload, expiresIn } = temporaryTokenTestCases.tampered;
        const token = CryptoUtils.generateTempToken(payload, parseInt(expiresIn));
        const tamperedToken = token.split('.')[0] + '.' + Buffer.from(JSON.stringify({hacked: true})).toString('base64') + '.' + token.split('.')[2];
        const result = CryptoUtils.verifyTempToken(tamperedToken);
        expect(result.valid).toBe(false);
    });
  
    it('should fail verification for an expired token', () => {
        const { payload, expiresIn } = temporaryTokenTestCases.expired;
        const token = CryptoUtils.generateTempToken(payload, parseInt(expiresIn));
        const result = CryptoUtils.verifyTempToken(token);
        expect(result.valid).toBe(false);
    });
  });

  describe('Other Utilities', () => {
    it('should generate a random salt', () => {
      const salt = CryptoUtils.generateSalt();
      expect(salt).toHaveLength(32);
      expect(typeof salt).toBe('string');
    });

    it('should generate a UUID', () => {
      const id = CryptoUtils.generateId();
      expect(id).toMatch(/^[a-f\d]{8}-[a-f\d]{4}-[a-f\d]{4}-[a-f\d]{4}-[a-f\d]{12}$/i);
    });

    it('should mask sensitive data', () => {
        const sensitive = '1234567890';
        expect(CryptoUtils.maskSensitiveData(sensitive)).toBe('1234**7890');
        expect(CryptoUtils.maskSensitiveData(sensitive, 2)).toBe('12******90');
        expect(CryptoUtils.maskSensitiveData('abc')).toBe('***');
    });
  });
}); 