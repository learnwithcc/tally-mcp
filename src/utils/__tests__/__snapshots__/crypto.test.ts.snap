// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoUtils API Key Generation and Validation should generate a valid API key with a custom prefix 1`] = `"testcf82da1a8437d38fb461a1482a1319c5a823a0715dec412ef85cefe7ea501f51_b07a9f69"`;

exports[`CryptoUtils API Key Generation and Validation should generate a valid API key with default options 1`] = `"tally_1401c0ca6a7c405a66811a6243236e06c748614a434524b9eb019b8d9a6a9cf4_f15c4314"`;

exports[`CryptoUtils Encryption and Decryption should encrypt and decrypt data successfully 1`] = `
{
  "encrypted": "c3e93b5340d3885370088ff865a554ebb6169a861698b74433",
  "iv": "e56edf67562d40e1a8d909955dd89e70",
  "tag": "c2216f58b5c21197e7fa124d311db2b1",
}
`;

exports[`CryptoUtils HMAC Signature should create and verify an HMAC signature 1`] = `"bf69d4f71ce097ea5010b1c69dbb5c43a2c00f125b07f1c4ad6740509ca1397b"`;

exports[`CryptoUtils Hashing should hash an API key consistently 1`] = `"682f6697fa93eca6c801a232519a09e3fe0c5c339465ee53c3f9edf92fd01f35"`;
